codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/EndpointGenerator.java

```diff
@@ -243,24 +246,39 @@ ObjectNode getService() {
             return regions;
         }
 
-        ArrayNode getVariants() {
+        private ArrayNode getVariants(ArrayNode mergedVariants) {
             List<Node> allVariants = new ArrayList<Node>();
 
             allVariants.add(getDefaultVariant(hostnameTemplate));
-            if (defaults.containsMember("variants")) {
-                ArrayNode variants = defaults.expectArrayMember("variants");
-                variants.forEach(variant -> {
-                    ObjectNode variantNode = variant.expectObjectNode();
-                    String hostname = variantNode.expectStringMember("hostname").getValue();
-                    String dnsSuffix = variantNode.getStringMemberOrDefault("dnsSuffix", this.dnsSuffix);
-                    String resolvedHostname = getResolvedHostname(hostname, dnsSuffix, endpointPrefix);
-                    allVariants.add(variantNode.withMember("hostname", resolvedHostname).withoutMember("dnsSuffix"));
-                });
-            }
+            mergedVariants.forEach(mergedVariant -> {
+                ObjectNode variantNode = mergedVariant.expectObjectNode();
+                String hostname = variantNode.expectStringMember("hostname").getValue();
+                String dnsSuffix = variantNode.getStringMemberOrDefault("dnsSuffix", this.dnsSuffix);
+                String resolvedHostname = getResolvedHostname(hostname, dnsSuffix, endpointPrefix);
+                allVariants.add(variantNode.withMember("hostname", resolvedHostname).withoutMember("dnsSuffix"));
+            });
 
             return ArrayNode.fromNodes(allVariants);
         }
 
+        private ArrayNode getMergedVariants(ArrayNode partitionVariants, ArrayNode serviceVariants) {
+            List<Node> mergedVariants = new ArrayList<Node>();
+
+            partitionVariants.forEach(partitionVariant -> {
```

Comment:
Does every partition's default contain every variant? Otherwise some service variants may lose.

Reply:
> Does every partition's default contain every variant?

Yes. The partition defaults contain every variant - to ensure that SDK has some endpoint to resolve.